File: app/src/main/cpp/wn-steam-client/src/pb/cpublishedfile.cpp
#include "wn_steam/pb/cpublishedfile.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

namespace {

// PublishedFileDetails is a large message (75+ fields) and only a handful are
// read here. Every case guards the tag's wire type before a typed read: a
// field whose wire type does not match the schema is skipped rather than read,
// so a single schema drift degrades gracefully instead of desyncing the
// cursor and corrupting the whole message.
using proto::WireType;

}  // namespace

std::vector<uint8_t> CPublishedFile_GetUserFiles_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.fixed64_field(1, steamid);
    w.uint32_field(2, appid);
    w.uint32_field(4, page);
    w.uint32_field(5, numperpage);
    w.string_field(6, type);
    w.uint32_field(14, filetype);
    return out;
}

std::optional<PublishedFileDetails>
PublishedFileDetails::parse(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    PublishedFileDetails m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        bool handled = false;
        switch (t->field_number) {
            case 1:  // result (varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u32(); v) m.result = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 2:  // publishedfileid (uint64 varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u64(); v) m.publishedfileid = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 5:  // consumer_appid (varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u32(); v) m.consumer_appid = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 7:  // filename (string)
                if (t->wire_type == WireType::LengthDelimited) {
                    if (auto v = r.string(); v) m.filename = std::move(*v);
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 8:  // file_size (uint64 varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u64(); v) m.file_size = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 10:  // file_url (string)
                if (t->wire_type == WireType::LengthDelimited) {
                    if (auto v = r.string(); v) m.file_url = std::move(*v);
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 11:  // preview_url (string)
                if (t->wire_type == WireType::LengthDelimited) {
                    if (auto v = r.string(); v) m.preview_url = std::move(*v);
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 14:  // hcontent_file (fixed64)
                if (t->wire_type == WireType::Fixed64) {
                    if (auto v = r.fixed64(); v) m.hcontent_file = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 16:  // title (string)
                if (t->wire_type == WireType::LengthDelimited) {
                    if (auto v = r.string(); v) m.title = std::move(*v);
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 20:  // time_updated (varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u32(); v) m.time_updated = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            default:
                break;
        }
        if (!handled) {
            if (!r.skip(t->wire_type)) return std::nullopt;
        }
    }
    return m;
}

std::optional<CPublishedFile_GetUserFiles_Response>
CPublishedFile_GetUserFiles_Response::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CPublishedFile_GetUserFiles_Response m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        bool handled = false;
        switch (t->field_number) {
            case 1:  // total (varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u32(); v) m.total = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 2:  // startindex (varint)
                if (t->wire_type == WireType::Varint) {
                    if (auto v = r.u32(); v) m.startindex = *v;
                    else return std::nullopt;
                    handled = true;
                }
                break;
            case 3:  // publishedfiledetails (repeated message)
                if (t->wire_type == WireType::LengthDelimited) {
                    auto b = r.bytes();
                    if (!b) return std::nullopt;
                    auto d = PublishedFileDetails::parse(*b);
                    if (!d) return std::nullopt;
                    m.publishedfiledetails.push_back(std::move(*d));
                    handled = true;
                }
                break;
            default:
                break;
        }
        if (!handled) {
            if (!r.skip(t->wire_type)) return std::nullopt;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cpublishedfile.rs
use crate::proto_wire::{Reader, WireType, Writer};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CPublishedFileGetUserFilesRequest {
    pub steamid: u64,
    pub appid: u32,
    pub page: u32,
    pub numperpage: u32,
    pub request_type: String,
    pub filetype: u32,
}

impl Default for CPublishedFileGetUserFilesRequest {
    fn default() -> Self {
        Self {
            steamid: 0,
            appid: 0,
            page: 1,
            numperpage: 1,
            request_type: String::new(),
            filetype: 0,
        }
    }
}

impl CPublishedFileGetUserFilesRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.fixed64_field(1, self.steamid);
        w.uint32_field(2, self.appid);
        w.uint32_field(4, self.page);
        w.uint32_field(5, self.numperpage);
        w.string_field(6, &self.request_type);
        w.uint32_field(14, self.filetype);
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct PublishedFileDetails {
    pub result: u32,
    pub publishedfileid: u64,
    pub consumer_appid: u32,
    pub filename: String,
    pub file_size: u64,
    pub file_url: String,
    pub preview_url: String,
    pub hcontent_file: u64,
    pub title: String,
    pub time_updated: u32,
}

impl PublishedFileDetails {
    pub fn parse(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            let handled = match (tag.field_number, tag.wire_type) {
                (1, WireType::Varint) => {
                    msg.result = reader.u32()?;
                    true
                }
                (2, WireType::Varint) => {
                    msg.publishedfileid = reader.u64()?;
                    true
                }
                (5, WireType::Varint) => {
                    msg.consumer_appid = reader.u32()?;
                    true
                }
                (7, WireType::LengthDelimited) => {
                    msg.filename = reader.string()?;
                    true
                }
                (8, WireType::Varint) => {
                    msg.file_size = reader.u64()?;
                    true
                }
                (10, WireType::LengthDelimited) => {
                    msg.file_url = reader.string()?;
                    true
                }
                (11, WireType::LengthDelimited) => {
                    msg.preview_url = reader.string()?;
                    true
                }
                (14, WireType::Fixed64) => {
                    msg.hcontent_file = reader.fixed64()?;
                    true
                }
                (16, WireType::LengthDelimited) => {
                    msg.title = reader.string()?;
                    true
                }
                (20, WireType::Varint) => {
                    msg.time_updated = reader.u32()?;
                    true
                }
                _ => false,
            };
            if !handled && !reader.skip(tag.wire_type) {
                return None;
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct CPublishedFileGetUserFilesResponse {
    pub total: u32,
    pub startindex: u32,
    pub publishedfiledetails: Vec<PublishedFileDetails>,
}

impl CPublishedFileGetUserFilesResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            let handled = match (tag.field_number, tag.wire_type) {
                (1, WireType::Varint) => {
                    msg.total = reader.u32()?;
                    true
                }
                (2, WireType::Varint) => {
                    msg.startindex = reader.u32()?;
                    true
                }
                (3, WireType::LengthDelimited) => {
                    msg.publishedfiledetails
                        .push(PublishedFileDetails::parse(reader.bytes()?)?);
                    true
                }
                _ => false,
            };
            if !handled && !reader.skip(tag.wire_type) {
                return None;
            }
        }
        Some(msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_user_files_response_and_skips_schema_drift() {
        let mut detail = Vec::new();
        {
            let mut w = Writer::new(&mut detail);
            w.uint32_field(1, 1);
            w.uint64_field(2, 99);
            w.uint32_field(5, 480);
            w.string_field(7, "file.bin");
            w.uint64_field(8, 1024);
            w.string_field(10, "https://example/file");
            w.string_field(11, "https://example/preview");
            w.fixed64_field(14, 777);
            w.string_field(16, "Workshop Item");
            w.uint32_field(20, 123456);
            w.fixed64_field(5, 0xfeed);
        }

        let mut body = Vec::new();
        {
            let mut w = Writer::new(&mut body);
            w.uint32_field(1, 1);
            w.uint32_field(2, 0);
            w.submessage_field(3, &detail);
        }

        let parsed = CPublishedFileGetUserFilesResponse::deserialize(&body).unwrap();
        let detail = &parsed.publishedfiledetails[0];
        assert_eq!(parsed.total, 1);
        assert_eq!(detail.publishedfileid, 99);
        assert_eq!(detail.consumer_appid, 480);
        assert_eq!(detail.hcontent_file, 777);
        assert_eq!(detail.title, "Workshop Item");
    }
}
